rust/kernel/lib.rs

```diff
@@ -70,6 +76,34 @@ pub trait Module: Sized + Sync {
     fn init(module: &'static ThisModule) -> error::Result<Self>;
 }
 
+/// A module that is pinned and initialised in-place.
+pub trait InPlaceModule: Sync + Send {
+    /// The type that implements the pinned initialisation of the module.
+    type Init: init::PinInit<Self, error::Error>;
+
+    /// Creates an initialiser for the module.
+    ///
+    /// It is called when the module is loaded.
+    fn init(module: &'static ThisModule) -> Self::Init;
+}
+
+impl<T: Module> InPlaceModule for T {
+    type Init = impl init::PinInit<Self, error::Error>;
+
+    fn init(module: &'static ThisModule) -> Self::Init {
```

Comment:
How about implementing this as a default implementation in `Module` other than the new `InPlaceModule`?

I.e. 
```rust
pub trait Module: Sized + Send + Sync {
    fn pin_init(module: &'static ThisModule) -> impl init::PinInit<Self, error::Error> {
        ...
    }
}
```

Of course, this requires RPITIT, but IIUC, it's going to be stabilized soon: https://github.com/rust-lang/rust/pull/115822

Reply:
Ah, I see, in the sample, `impl`s may be overwritten by a particular module, so it makes sense to have a different `trait`. But still you can use RPITIT ;-)